src/ca65/instr.c

```diff
@@ -373,6 +381,153 @@ static const struct {
     }
 };
 
+/* Instruction table for the Rockwell C39 */
+static const struct {
+  unsigned Count;
+  InsDesc  Ins[135]; // FIX WHEN ADDING INSN!
+} InsTabC39 = {
+    sizeof (InsTabC39.Ins) / sizeof (InsTabC39.Ins[0]),
+    {
+      // C39: (ind)   = AM65_DIR_IND   = 0x0400 allowed
+      // C39: (ind,x) = AM65_DIR_X_IND = 0x8000 not allowed
+      { "ADC",  0x080266C, 0x60, 0, C39PutArith }, 
+      { "ADX",  0x0800024, 0x80, 0, C39PutAdd }, // C39: renamed ADD to ADX because it conflicts with add macro
```

Comment:
The Rockwell ADD insn has only 3 AMs: IMD, ZP and ZP,X so renaming the
add-macro will not work.

ADD Add Memory to Accumulator without Carry

Operation: A + M → A, C N V – B D I Z C

C C – – – – C C

ADD adds the value in memory to the value in the accumulator and
stores the result in the accumulator.  The C flag is set when the sum
exceeds hexadecimal 255 (binary mode) or decimal 99 (decimal mode),
otherwise C is reset.  In the binary mode, the V flag is set when the
sign (bit 7) changes due to the result exceeding +127 or –128,
otherwise V is reset. In the decimal mode, V is reset. The N flag is
set if bit 7 of the result is a 1, otherwise N is reset. The Z flag is
set if the result is zero, otherwise Z is reset.

Addressing Assembly Language Op No. No.
Mode Form Code Bytes Cycles Ref.

Immediate ADD #Oper 89 2 2c C2.1
Zero Page ADD Oper 64 2 3c C3.1
Zero Page,X ADD Oper,X 74 2 4c C4.1

c. Add 1 if in decimal mode.



On 08.05.2015 05:25, Christian Groessler wrote:
>
> In src/ca65/instr.c 
> <https://github.com/cc65/cc65/pull/157#discussion_r29914926>:
>
> > @@ -373,6 +381,153 @@ static const struct {
> >      }
> >  };
> >
> > +/* Instruction table for the Rockwell C39 */
> > +static const struct {
> > +  unsigned Count;
> > +  InsDesc  Ins[135]; // FIX WHEN ADDING INSN!
> > +} InsTabC39 = {
> > +    sizeof (InsTabC39.Ins) / sizeof (InsTabC39.Ins[0]),
> > +    {
> > +      // C39: (ind)   = AM65_DIR_IND   = 0x0400 allowed
> > +      // C39: (ind,x) = AM65_DIR_X_IND = 0x8000 not allowed
> > +      { "ADC",  0x080266C, 0x60, 0, C39PutArith },
> > +      { "ADX",  0x0800024, 0x80, 0, C39PutAdd }, // C39: renamed ADD to ADX because it conflicts with add macro
>
> I think renaming ADD is not a good idea if it's called ADD in the 
> official C39 documentation. The ADD macro should be renamed.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/cc65/cc65/pull/157/files#r29914926>.
>

-- 
Mit freundlichen Grüssen

Rainer Keuchel